src/main/java/com/google/testing/junit/testparameterinjector/ParameterWrapper.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2021 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.testing.junit.testparameterinjector;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Method;
+import java.lang.reflect.Parameter;
+import javax.annotation.Nullable;
+
+abstract class ParameterWrapper {
+  @SuppressWarnings("AndroidJdkLibsChecker") // j.l.r.Parameter is not available on old Android SDKs.
+  static ParameterWrapper[] get(Constructor<?> constructor) {
+    try {
+      return Java8.create(constructor.getParameters());
+    } catch (NoSuchMethodError ignored) {
+      return Legacy.create(constructor);
+    }
+  }
+
+  @SuppressWarnings("AndroidJdkLibsChecker") // j.l.r.Parameter is not available on old Android SDKs.
+  static ParameterWrapper[] get(Method method) {
+    try {
+      return Java8.create(method.getParameters());
+    } catch (NoSuchMethodError ignored) {
+      return Legacy.create(method);
```

Comment:
Why not always use the Legacy version?

Reply:
The `Java8` wrapper correctly propagates parameter names (assuming they're available in the class file). I did not look to see how the name is used in practice, though.

Reply:
The parameter name is to (optionally) format the name of the test:

```
myTest[myParameter=true]
```

In any case, I've been investigating whether we can use `AnnotationWithMetadata` a bit earlier in the process. I'm fine with merging this PR for now and see if I can improve on it at a later point.